types/src/nmt.rs

```diff
@@ -18,34 +37,128 @@ pub use self::namespaced_hash::{
+///
+/// [`NamespacedHash`] can represent the leaves or internal nodes of the [`Nmt`].
+///
+/// If it is a hash of a leaf, then it will be constructed like so:
+///
+/// `leaf_namespace | leaf_namespace | Sha256(0x00 | leaf_namespace | data)`
+///
+/// If it is a hash of the children nodes, then it will be constructed like so:
+///
+/// `min_namespace | max_namespace | Sha256(0x01 | left | right)`
 pub type NamespacedHash = nmt_rs::NamespacedHash<NS_SIZE>;
+/// Hasher for generating [`Namespace`] aware hashes.
+pub type NamespacedSha2Hasher = nmt_rs::NamespacedSha2Hasher<NS_SIZE>;
+/// [`Namespace`] aware merkle tree.
 pub type Nmt = nmt_rs::NamespaceMerkleTree<MemDb<NamespacedHash>, NamespacedSha2Hasher, NS_SIZE>;
 
-pub struct NodePair(NamespacedHash, NamespacedHash);
-
+/// Namespace of the data published to the celestia network.
+///
+/// The [`Namespace`] is a single byte defining the version
+/// followed by 28 bytes specifying concrete ID of the namespace.
+///
+/// Currently there are two versions of namespaces:
+///  - version `0` - the one allowing for the custom namespace ids. It requires an id to start
+///  with 18 `0x00` bytes followed by a user specified suffix (except reserved ones, see below).
+///  - version `255` - for secondary reserved namespaces. It requires an id to start with 27
+///  `0xff` bytes followed by a single byte indicating the id.
+///
+/// Some namespaces are reserved for the block creation purposes and cannot be used
+/// when submitting the blobs to celestia. Those fall into one of the two categories:
+///  - primary reserved namespaces - those use version `0` and have id lower or equal to `0xff`
+///  so they are always placed in blocks before user-submitted data.
+///  - secondary reserved namespaces - those use version `0xff` so they are always placed after
+///  user-submitted data.
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd, Hash)]
 pub struct Namespace(nmt_rs::NamespaceId<NS_SIZE>);
 
 impl Namespace {
+    /// Primary reserved [`Namespace`] for the compact [`Share`]s with [`cosmos SDK`] transactions.
+    ///
+    /// [`Share`]: crate::share::Share
+    /// [`cosmos SDK`]: https://docs.cosmos.network/v0.46/core/transactions.html
     pub const TRANSACTION: Namespace = Namespace::const_v0([0, 0, 0, 0, 0, 0, 0, 0, 0, 1]);
+
+    /// Primary reserved [`Namespace`] for the compact [`Share`]s with [`MsgPayForBlobs`] transactions.
+    ///
+    /// [`Share`]: crate::share::Share
+    /// [`MsgPayForBlobs`]: celestia_proto::celestia::blob::v1::MsgPayForBlobs
     pub const PAY_FOR_BLOB: Namespace = Namespace::const_v0([0, 0, 0, 0, 0, 0, 0, 0, 0, 1]);
+
+    /// Primary reserved [`Namespace`] for the [`Share`]s used for padding.
+    ///
+    /// It is used for [`Share`]s inserted after all other shares from primary reserved namespace
+    /// to align user-defined namespaces correctly in [`ExtendedDataSquare`].
```

Comment:
```suggestion
    /// [Share]s with this namespace are inserted after other shares from primary reserved namespace
    /// so that user-defined namespaces are correctly aligned in [`ExtendedDataSquare`]
```
idk if it's me, but I find original hard to parse - feel free to include or not

Reply:
sure, just corrected backticks